Recurrent Attention API Additions

andhus opened the pull request:
This PR is the first step to support recurrent attention mechanisms as suggested in https://github.com/fchollet/keras/issues/7633.

In short, it adds the possibility to pass `constants` (besides `inputs` and `initial_state`) when calling the RNN layer. The RNN will pass on the tensor(s) to the call method of the RNN  Cell which (_if_ constants are passed) is expected to implement the signature:
```call(inputs, states, constants)```

roya0045 commented:
Any idea why it seems to fail with CNTK (according to Travis)?

fchollet commented:
Can you explain why this would be necessary? I believe RNN cells that need specific constants should be managing their own constants, e.g. what we do for dropout (the parent RNN is agnostic to dropout constants).

I am pretty sure Attention wrappers would be implementable with the current API

andhus commented:
The key here is that the constants (the “attended” when used for attention) _really are inputs_ to the the recurrent layer (I refer to them as “external” contacts in docs but maybe we need a better name). I think it would be bad and confusing to break the principle that _all inputs to a layer should be passed to that layer_ in `__call__`, i.e. when it is _applied_ to its inputs in the functional API.

The constant(s) needed for attention is fundamentally different from dropout, as the later is only a regularisation method and we _don’t pass the dropout values in inference mode_ (they are only generated in in training) but the attended of course have to passed as in input.

When it comes to implementation/API, it is certainly possible to implement attention in keras only using custom wrapper - I’ve done it several times and that was the solution in the original API suggestion, which I think is much worse  ;).

But let’s looks at some options...

This is the “look and feel” to the user of my current suggestion:

```
input_sequence = Input((timesteps, features))
attended_sequence = Input((attended_timesteps, attended_features))

attention_lstm = RNN(
    cell=SequenceAttention(LSTMCell(32)),
    return_sequences=True  # could of course be False in another use-case
)
lstm_output_sequence = attention_lstm(
    input_sequence,
    constants=attended_sequence
)

output_layer = TimeDistributed(Dense(n_labels, activation='softmax'))
output = output_layer(lstm_output_sequence)

model = Model(inputs=[input_sequence, attended_sequence], outputs=output
```

I guess what your suggestion means is that you would only let the cell know about the attended, something like:

```
cell=SequenceAttention(LSTMCell(32), attended=attended_sequence)
attention_lstm = RNN(cell=cell, return_sequences=True)
lstm_output_sequence = attention_lstm(constants=attended_sequence)
```

Or maybe:

```
cell=SequenceAttention(LSTMCell(32))
cell.attended=attended_sequence
attention_lstm = RNN(cell=cell, return_sequences=True)
```

You can probably make this work but I'd like to avoid it for a number of reasons: 

1. I don’t think there is any other situation in keras where one would _inject an input tensor in the init of a layer_ - it feels intuitively wrong to have these implicit inputs (for which we, again, have to pass data for both in training an inference). 

2. If not all inputs are passed in call, it breaks the possibility to easily reuse the layer in several places of the computational graph.

3. Typically you need to _build_ the attention mechanism (initialise weights etc.) and I see no reason why this shouldn’t be done in the build method (triggered by `__call__`) where you get the input shapes of all inputs.

I think a good comparison is why we can (optionally) pass initial state to the RNN. If we do it should really be considered an input (needs to be passed in inference) - otherwise it's an "internal thing" like dropout. Just like the initial state, the attended can (or _will_ in most cases) be the output of a previous layer.

Another option would be to treat the attended as part of the `initial_state` but that's a hack a think and leads to several challenges.

Yet another (better) option is to add a new class `AttentionRNN(RNN)` which adds the functionality suggested in this PR without touching the existing RNN - but then you need to both wrap the cell with an attention mechanism and use another Parent RNN layer for using it. This would be fine I think but since the addition in `RNN` just adds an additional optional argument and is backward compatible I thought it would be better this way.

It is also, of course, possible that I'm missing some other option. Please question the above!

andhus commented:
The fact that we in the current RNN implementation "temporarily" modify the `input_spec` when `initial_state` is passed is very related to this discussion: https://github.com/fchollet/keras/blob/master/keras/layers/recurrent.py#L453
I've just extended the same behaviour for constants but I don't really like it. Just like  @farizrahman4u points out in a comment on the second PR: https://github.com/fchollet/keras/pull/7981 it would be better to be explicit about all tensors that really are inputs in the `input_spec`

fchollet commented:
Thanks for the detailed explanation! It's a reasonable choice. An alternative would be the following:

- `AttentionRNN` subclass of RNN provides support for attention
- `attended_sequence` keyword argument in the `call` of `AttentionRNN`
- which gets added to the input list (just like we do for the `initial_state` argument)

```python
y = AttentionRNN(cell)(x, attended_sequence=z)
```

What do you think?

In terms of PR process, I suggest doing a single PR for the whole thing. That will make the purpose of the various pieces more apparent.

andhus commented:
Sounds great! In summary, the additions to `RNN` in this PR are kept separate in a new subclass class `AttentionRNN`. I suspect this will introduce code duplication but I agree this is motivated by avoiding any additional complexity to the `RNN`.

A few details:
- The new input keyword argument should not contain `sequence` as it could be just anything, e.g. an image in an "mage captioning problem". What really defines this argument is that it will be presented in "whole" at each cell transformation. How about just `attended`?. It's required shape is only defined by cell implementation, for attention-cells that implement `attended_spec` the `AttentionRNN` can expose this (including the number of dims already before application).

- A reminder (that is relevant for the naming discussion) is that there is no reason to limit ourselves to a single attended input (let's say you want to train a model to summarise a news article and attend both to the text and an image e.g.). I think the plural covered by `attended` but we could also go for `attended_inputs`.

- Just to be clear, we'll need to add the new keyword to `__call__` as well as we need to override parent `RNN.__call__` to add the appended to `input_spec` before build.

So in summary:
```
y = AttentionRNN(attention_cell)(x, attended=z)

# alt. for many attended:

y = AttentionRNN(attention_cell)(x, attended=[z1, z2])

```
Totally agree about the PRs, Let's close second one to keep discussion in one place!

fchollet commented:
Cool, let's go with this then!

andhus commented:
Great, I'll update this PR and the API doc this weekend.

andhus commented on keras/layers/recurrent.py:
```diff
@@ -194,6 +196,130 @@ def get_losses_for(self, inputs=None):
         return losses
 
 
+class FunctionalRNNCell(Wrapper):
```
Is `RNNCellModel`better name?

andhus replied:
NO, that's not a good idea. It's not a Model in the sense that it should not have `compile`, `fit`, `predict` etc... it's should just "package" the cell based on the functional definition.

andhus replied:
We should probably override `__call__` as well.

fchollet commented:
This is a big PR. If any users of Attention or people familiar with Keras RNNs could help with reviewing, that would be greatly appreciated.

andhus commented:
Yes, would be great with input from more contributors/users. Here is a guide/overview of the additions:

*New class* `FunctionalRNNCell`
Adds support for composing RNN cells using keras functional API. (Alternative name could be RNNCellModel?). 
- Compared to the general `Model` it automatically implements the property `state_size` as well as the correct signature of `call` (required for RNN cells) based on the arguments passed.
- It also offers a better API for keeping track of what is input/output and what is input _states_/output _states_.
- Supports an additional input `attended` for implementing attention mechanisms.

*New class* `AttentionRNN(RNN)`
There is quite a lot of code here but it really just does a few things 
- Asserts that the passed cell supports attention (`cell.call` takes kwarg `attended`)
- Takes the kwarg `attended` (tensor) in both `__call__` and `call` and passes it on to the cell.
- Temporarily modifies `input_spec` in `__call__` so that the `attended`is properly added to the computational graph.
- _If_ the wrapped cell is a keras `Layer` it passes the shape of the `attended` as part of `input_shape` to `cell.build` so that the cell can build the attention mechanism properly.

In the `AttentionRNN` there is quite some code duplication from `RNN`(as expected). This can be alleviated by breaking apart the `__call__` and `call`method of `RNN` but base on our previous discussions I intentionally left `RNN` untouched now.

There are also many more tests required, but I want some feedback/review before continuing.

fchollet commented:
> New class AttentionRNN(RNN)

Could we move attention support to the base class `RNN` altogether? Conditional on whether the cell's `call` takes the `attended` argument.

Also: what would be some alternative names for `attended`? `attended_sequence`? I'm afraid `attended` wouldn't be sufficiently explicit.

andhus commented:
> Could we move attention support to the base class RNN altogether?
This is _exactly_ what I implemented in a backward compatible way (and tested) in the original version of this PR (i then used the keyword argument `constants`): https://github.com/fchollet/keras/pull/7980/commits/28795f14ba734d0d97cf292e601d7ab00039f401

> Also: what would be some alternative names for `attended`? `attended_sequence`?
As discussed a few comments above, "sequence" should not be in the name. Again, the attended can be anything (and multiple) tensors. A standard application is image captioning where the attended is an image (or typically CNN feature maps of an image).

I do still think it would be better to support passing, what I originally referred to i docs as, _external constants_ (in lack of better name) to the base `RNN` which forwards them to the `cell.call`, as well as their shapes to `cell.build` if the cell is a keras layer. And yes, if such constants are passed in `RNN.__call__` it is (naturally) required that  `cell.call` accepts them. (Note: I really recommend that we use a specific keyword argument for this in the cell and don't append constants to the states as was done for "dropout constants" before, for a number of reasons out of this scope).

As per the current state of this PR `RNN` is purely a special case of `AttentionRNN` (with `attended=None`) which is a bit strange. I agreed to your suggestion to put this logic in a new separate class with (as I wrote) the main motivation being "avoiding adding any additional complexity to the `RNN`" and perhaps to move forward a bit faster ;). There is value in minimising complexity of the base `RNN`class, for example for semi-advanced users that want to extend it, but the question is if this motivates a whole new class which is just a _generalisation_ of the first.

Regarding naming: If we were to stick to an additional separate `AttentionRNN` I think `attended` is the best name. As suggested previously, `attended_inputs` is also an alternative. However, if we move back support of "external constants" to `RNN` I think `attended` is a confusing and unnecessary _limiting_ name. I can't come up with a better name than `constants` for this case, and make clear in documentation that "these are tensors that will be passed _in whole_ to `cell.call` for each time step" as was done in original PR, see: https://github.com/fchollet/keras/pull/7980/commits/28795f14ba734d0d97cf292e601d7ab00039f401#diff-3118e4e28157032506f771f279a551c3R281

I'd prefer to consider `RNN` the component "in the keras `Layer`-abstraction level" that corresponds to the keras backend's `K.rnn` or e.g. `theano.scan`. There is a reason why both of these takes an argument for "external constants". In `K.rnn` the keyword `constants` is already used and in `theano.scan` uses `non_sequences`. In theano there are important optimisation reasons for passing these (not utilised in keras backend!?) and it is not required to pass such constants to a "control flow op" if it suports "partialising" of the step function with preset tensors. We discussed this "partialisation" in our initial discussion and it is not an option on the keras `Layer`-abstraction level as we need (want!) to be explicit about all inputs and outputs of layers.

The theano "non_sequences" is not a very user-friendly name as the "external constants" could be sequences _but not iterated over by the RNN cell_. The reason I don't want to go for `attended`as keyword if support is added ot `RNN` is that they could be used for other things (you could e.g. write cell wrappers that implements recurrent dropout for any cell based on these constant...)


...I apologise for these very long comments. This is an intricate matter. I've really tried to do a thorough requirements analysis with respect to both how Attention is used in research and the current state of Keras (which e.g. lead to my conclusion that the RNNCell should really be broken out before this addition). Perhaps it was too early to move to PRs but on the other hand it is the only way to really show the implications and feasibility of the suggestion. It is quite slow to work with the sparse, and occasionally inconsistent ;), feedback. I'd be happy to set up a Hangouts session to discuss details and options directly!

andhus commented:
Summary of the very long comment above:
- I think best option is to add support for (non iterated over) constants in `RNN` as per the original solution in this PR: https://github.com/fchollet/keras/commit/28795f14ba734d0d97cf292e601d7ab00039f401 This way we keep down the umber of classes, and for implementing attention we just need to wrap/implement the RNNCell.
- In this case I can't come up with a better name then `constants`. It can be made clear from docs that these are tensors passed "in whole" to `cell.call`, i.e. they are "not iterated over". I'd prefer to use such a general name because it could be used for other things then "attention" - however if we think it will only be used for attention we could name the argument `attended` or `attended_inputs`.
- It would be great with feedback from more contributors. If we want to speed up the process I'd be happy to set up a hangouts session for discussion. I've already explored the pros and cons of many alternatives but its challenging to present the full analysis in comments here...

fchollet commented:
> I think best option is to add support for (non iterated over) constants in RNN as per the original solution in this PR: 28795f1 This way we keep down the number of classes, and for implementing attention we just need to wrap/implement the RNN cell.

I've been looking at these PRs for some time and I'm still not sure. It would definitely be helpful to have more people chime in with their perspective.

I think you're probably right, it would be good to add `constants` to the `RNN` API. One thing to consider though, is that `keras.backend.rnn` already has support for `constants`. It would be good not to reimplement an independent, layer-level mechanism, and instead reuse what we have...

If we do this, the API changes would be:

- `constants=None` in `RNN`'s `call`. Ideally this would forward the constants to `K.rnn` (they are then appended to the first argument passed to the step function).
- `SequenceAttention` wrapper class for RNN cells? (please confirm)

> In this case, I can't come up with a better name then constants

`constants` is fine.

fchollet commented:
> It would be good not to reimplement an independent, layer-level mechanism, and instead reuse what we have...

Besides just the semantic problem of having multiple similar but independent mechanisms with the same name, one thing to consider is performance. Different backends might want to optimize the way they manage constants. Theano does it to some extent, for instance.

fchollet commented:
Let's see... how bad would it be to have something like:

```python
def __call__(self, inputs, initial_state=None, constants=None, **kwargs):
       # handle connectivity of `constants`, like we do for `initial_state`
       ...

def call(self, inputs, mask=None, training=None, initial_state=None, constants=None):
        ...
        if constants:
            if not has_arg(self.cell.call, 'constants'):
                 raise ValueError(...)
            def step(inputs, states):
                 constants = inputs[1:]
                 if has_arg(self.cell.call, 'training'):
                     return self.cell.call(inputs[0], constants=constants, training=training)
                 else:
                     return self.cell.call(inputs[0], constants=constants)
           
        last_output, outputs, states = K.rnn(step,
                                             inputs,
                                             initial_state,
                                             go_backwards=self.go_backwards,
                                             mask=mask,
                                             constants=constants,
                                             unroll=self.unroll,
                                             input_length=timesteps)
```

andhus commented:
Good, at this time it seems obvious to me that we should add `constants` support to `layer.RNN`. As a "non-advanced user" you don't have to now that it's even there.
   
> It would be good not to reimplement an independent, layer-level mechanism, and instead reuse what we have...

I'm not sure exactly what mean by this, but yes it's important that `Layer` level API is consistent (in terminology and functionality) with the `backend` level to the extent possible/meaningful. It is e.g. a good motivation to stick to the same name and behaviour for `constants`.

> Besides just the semantic problem of having multiple similar but independent mechanisms with the same name [...]

I don't think there is anything strange about having a layer `RNN` that to a great extent "duplicates" functionality of backend `rnn`, because I think it is really clear what the layer version "brings to the table" (in terms of initialisation, building, holding weights...). Backend `rnn` is the purer (subset) of `RNN` functionality and responsibility that is backend dependent.

> [...] one thing to consider is performance. Different backends might want to optimize the way they manage constants. Theano does it to some extent, for instance.

Yes, this is important and quite complex. I wrote about it in my previous comments and put a note on it in the PR (https://github.com/fchollet/keras/pull/7980/files#diff-3118e4e28157032506f771f279a551c3R2340). I agree that its better to pass constants to backend `rnn` and let it explicitly pass them to the `step_function` (rather then partialising step_function with them preset), but regarding optimisation I'm not sure that is "enough" or makes a big difference anyway. For theano we should actually pass all involved tensors in the step transformation, _including the weights_ (!), as `non_sequences` to `theano.scan`. This is not done/supported with current API. I haven't profiled profiled how big difference it makes for typical applications.

> Let's see... how bad would it be to have something like: [...]

This is not too bad - it's one way of making sure that the constants are passed via backend `rnn`. There is a bug though: you assume `constants` are appended to `inputs` inside `rnn` but in fact they are appended to `states` (see e.g.: https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py#L2479). Note that this appending of constants is not documented in any of the backends `rnn`. CNTK has no docs and both Theano and Tensorflow backends are in fact wrong when stating:

>  `new_states: list of tensors, same length and shapes as 'states'`

Where `'states'` referens to the input of the `step_function`. This is not true if constants are passed to `rnn` as then this is the case: `len(new_states) < len(states)`.
   
I think this mistake is very symptomatic of some of the issues with the Keras recurrent code base - _that arguments are not always what they "claim to be" - and will be different things at different times_: "states can be either states - or states and constants" and "inputs can be either inputs - or inputs and initial_state" This causes mental overhead, boilerplate for splitting lists of inputs and decreased readability.

I'd really recommend working towards removing this pattern. We should try to achieve that the `step_function`in backend `rnn` reflects the RNNCell in `RNN` and I think there should really be a separate argument for constants and not append them to states as is done now (or inputs as would be required in your code example) - this gos also for the `step_function`. Everything becomes much clearer and easier to extend. This is super easy to fix but would not be backward compatible in terms of backend `rnn`'s expectations on step_function (although, as mentioned above, its not even part of API documentation how constants are passed). What's your thoughts on this? 

Another example is that `inputs` to `RNN` can be either just the input tensor or a list of the input _and_ `initial_state`. If we add yet another keyword argument `constants` to `RNN.__call__` that could also be passed  as part of `inputs` it gets even more complicated. If `inputs`is a list of say two tensors, how can we know it the second one is `intitial_state`or `constants`? Especially if `state_spec` and `constant_spec`happen to be the same? Is this really required, is it not possible to allow passing `initial_state` (and later `constants`)  _only_ by keyword arg? I get that is has to do with composing (and serialisation?) of models, I haven't looked into the details but there should be a better solution...

andhus commented:
And then to the actual topic of "attention" :D...

> - SequenceAttention wrapper class for RNN cells? (please confirm)

Yes, as long as we make sure `constants` are passed to `cell.call`, the only thing needed to implement/configure attention mechanism is a wrapper of the "core" RNN cells.

My suggestion is that we add at least one standard attention mechanism for sequences (batch dim: 3) and images (batch dim: 4) respectively.

In almost all applications of attention to spatial input, the mechanism consists of three steps:

1. Predict a _distribution over attended spatial dimension_ (based on state of core cell and input at current time step) - i.e. "decide where to look".
2. Multiply attended with this distribution and average over spatial dimensions to get a singel _attention feature vector_.
3. Pass the attention feature vector,  or concatenation of attention feature vector and original input as input to core cell (together with states).

The "flow" of these operations will be identical for any attention mechanism/input shape, and can be implemented in a `AttentionCellWrapper` base class (I have already done this).  The only thing that needs to vary for each type of attention mechanism is step 1. - predicting of a _distribution over input spatial dimension_. This is a very neat solution: we've reduced the problem of recurrent attention to _predicting a distribution_ (1D, 2D, ...).

andhus commented:
This means one very modular (and according to me beautiful) approach would be:

```
sequence_attention_rnn = RNN(
    DistributionAttention(
        LSTMCell(...),
        distribution=MixtureOfGaussian1D(...),
        ...
    ),
    ...
)
output = sequence_attention_rnn(input_sequence, constants=attended_sequence)
```

The only thing that would be changed for image attention (e.g. image captioning):

```
image_attention_rnn = RNN(
    DistributionAttention(
        LSTMCell(...),
        distribution=MixtureOfGaussian2D(...),  # NOTE 2D
        ...
    ),
    ...
)
output = image_attention_rnn(input_sequence, constants=image_features)
# image_features would typically be the output from a Conv. stack (4D tensor)
```

This is a truly modular, and I think also intuitive, way. But of course you can alos prepackage attention cell wrappers in a single class like `SequenceAttention` and `ImageAttention`.

A great thing about the approach suggested is that the same distributions can be used as output layers for mixture density networks and no duplication of core logic is needed for feed-forward attention (not so commonly used though). This is very feasible to implement, I'm happy elaborate and show examples...

andhus commented:
Here's an example of how the "distribution" can be implemented: https://github.com/andhus/extkeras/blob/master/extkeras/layers/distribution.py#L102
But maybe better to just implement prepackaged attention wrappers first.

Also, note that by adding the `FunctionalRNNCell`in this PR you can use functional API to compose any RNN cell and attention mechanism using existing or custom Keras layers.

fchollet commented:
Thanks for the explanations! The modular API you propose look good to me. So we should:

- Add the `constants` argument in RNN's `call` that you proposed. It should be implemented by passing `constants` to the `K.rnn` call.
- Add an Attention RNNCell wrapper.

Let's hold off on the model-like API for instantiating cells. It's not clear at the moment that this is preferable to simply writing your own RNNCell, so we should not commit to a dedicated API yet.

fchollet commented:
> one of the issues with the Keras recurrent code base - that arguments are not always what they "claim to be" - and will be different things at different times: 

The specific cases you list arise from one simple problem: topology management requires us to normalize all call signatures to a standard signature where all tensor inputs are in the first argument, as a flat list. However, good API design sometimes requires dedicated keyword arguments for some inputs (`initial_state`, `constants`). Hence the need to have a conversion layer between what's user-facing (the keyword args) and the internal representation (single flat list of inputs). This is purely internal code and it's fine to have it.

andhus commented:
Sure, I haven't looked into the details, but it would have been better with support for _named_ links in the topology based on naming of args. When I call:
```
lstm = LSTM(...)
y = lstm(x, initial_state=[h, c])
```
 I'd have expected `lstm.inputs` to return something like `{'inputs': <tf.Tensor ...>, 'initial_states': [<tf.Tensor ...>, <tf.Tensor ...>]}` and same structure for `lstm.inbound_nodes[0].inbound_layers`.

andhus commented:
To proceed, there are two things from past discussions I'd like to get confirmed:

1) Is it part of the (functional) API that you should be able to pass `initial_states` as part of `inputs` (instead of using the keyword argument) _first time_ the RNN layer is called? And do we want the same behaviour for the new argument `constants`?

For example should you, first time calling a LSTM, be able to do this:
```y = lstm([x, h_init, c_init])```,
and should you be able to, first time calling an attentive lstm, do this:
```y = att_lstm([x, z])```,
as well as: 
```y = att_lstm([x, h_init, c_init, z])```.
And for the case of multiple constants (attended):
```y = att_lstm([x, z1, z2])```,
as well as:
```y = att_lstm([x, h_init, c_init, z1, z2])```

If so, _inferring_ what `inputs` actually contains (in `RNN.__call__`) gets complicated. To support all corner cases, an attentive cell would have to _declare_ how many constants it expects (to make it possible for RNN to split inputs and know if it contains initial states as well) which is otherwise not needed. I emphasise _first time_ in my question because (I think) as long as you used the keyword argument `constants` first time calling the `RNN` you can store what's expected in e.g. a property `constants_spec` (default to `None`) to allow general splitting of inputs if called again. But it gets a bit messy anyhow (hence my "wish" in previous comment).

2) Can we change expected signature of `step_function` in backend `rnn` to:
```step_funtion(inputs, states, [constants])```
(where constants are only expected if constants are passed to `rnn`) to reflect how we want `RNNCell.call` signature?

I can't see any reason why we append `constants` to `states` inside `rnn`, this case is not motivated by "flat list" requirements from Topology linking. (If completely ignoring legacy, I'd rename `step_function -> cell_function` as well). However, I do realise that this breaks the expected (even if not documented) behaviour of `rnn`. If we don't change this, it will result in an inconsistency in how constants are handled in `RNN <-> cell` vs `rnn <-> step_function`. However, if this is not an option, _I think it is still better to accept the inconsistency rather than enforcing the "unmotivated appending of constants to states" in_ `RNNCell.call`.

andhus commented:
> _I think it is still better to accept this inconsistency, rather than enforcing the "unmotivated appending of constants to states" in_ `RNNCell.call`
Please trust me on this one; it will _destroy_ readability of Attention Mechanisms and cause a lot of implementation overhead when trying to make them modular if we start mixing `states` and `constants` in the same list.

fchollet commented:
Yes, passing all input tensors as a flat list is part of the internal API. On the user-facing side of things, we can always avoid doing that (and instead pass the additional arguments as keyword arguments), but it is unavoidable at model deserialization time.

I agree that it's difficult to figure out a reasonable way to parse the list of inputs. Since the problem only occurs at deserialization time, it should be possible to store in the layer config whether or not initial states and constants are "keras tensors" (i.e. whether they are part of the model topology), which would then provide all the information needed for deserialization.

Keep in mind that this is an implementation detail. The public API is still to use keyword arguments.

fchollet commented:
> Please trust me on this one; it will destroy readability of Attention Mechanisms 

It is purely an implementation detail and will not modify the signatures of the `call` methods of either batch-level layers or cells. Hence no readability problem. We should do it, both for performance reasons and conceptual simplicity. See the earlier code snippet for an implementation suggestion.

andhus commented:
> The public API is still to use keyword arguments

Ok great, (1.) is not a problem then. As you suggest, `RNN` can store what's expected (in e.g. `constants_spec`), then the splitting of internal flat-list-format will be quite trivial. And yes, this has to be represented in the config (in a backward compatible way).

I'm not sure I understood the answer to (2.), is my understanding correct that we should _not_ change expected signature of (backend `rnn` argument) `step_function`? But instead just wrap `cell.call` inside `RNN`?

andhus commented:
> It is purely an implementation detail [...] hence no readability problem.

I totally appreciate Keras extreme focus on the public API. But for the library (and API) to stay up to date it has to continue to develop. The structure and readability of its internals are crucial for easy extension and for people to continue to contribute. Many improvements could be done that are in no conflict with public API. ...The quote above scares me a bit ;)

fchollet commented:
> I'm not sure I understood the answer to (2.), is my understanding correct that we should not change expected signature of (backend rnn argument) step_function? But instead just wrap cell.call inside RNN?

Yes, indeed, we can simply wrap the calls to the step function, as in the code snippet.

Your other comment: my quote on readability means to say that since method signatures don't change, the implementation of `call` is strictly the same. Hence is it not less readable: it is identical.

andhus commented:
Ok, so I've updated this PR back to  "supporting of constants in `RNN`" according our discussion.

For overview, there is a functioning implementation (based on the additions in this PR) of next steps:
- base class for attention cell wrappers
- mixture of gaussian 1D/sequence attention (including example)
here: https://github.com/andhus/keras/pull/3. I didn't want to pull it in to this PR until we've settled on the changes in `RNN`.

There is also an explanation of the example in the linked PR (and a benchmark MoGAttention vs. regular seq-to-seq) here: https://github.com/andhus/extkeras/tree/master/examples/attention

andhus commented on keras/layers/recurrent.py:
```diff
@@ -363,13 +372,11 @@ def __init__(self, cell,
-                               for dim in self.cell.state_size]
-        else:
-            self.state_spec = InputSpec(shape=(None, self.cell.state_size))
+        self.state_spec = None
```
For more advanced cells (typically attention cells such as this one: https://github.com/andhus/keras/pull/3/files#diff-9c4188ddc4dd173d80f64feed5b89412R258) the `cell.state_size`is not defined until the cell is built. But that's not a problem, `state_spec`is just set a bit later in build (or in `__call__` if `initial_state`is passed to `RNN`)

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -440,62 +474,68 @@ def get_initial_state(self, inputs):
-            initial_state = inputs[1:]
-            inputs = inputs[0]
+    def __call__(self, inputs, initial_state=None, constants=None, **kwargs):
+        inputs, initial_state, constants = self._normalize_args(
```
Prefer "standardize" (also used elsewhere) since "normalize" has a specific meaning elsewhere.

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -440,62 +474,68 @@ def get_initial_state(self, inputs):
+            check_list += constants
+            self.constants_spec = [InputSpec(shape=K.int_shape(constant))
+                                   for constant in constants]
+            self._n_constants = len(constants)
```
In line with naming conventions in this API, this should be `_num_constants`.

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -440,62 +474,68 @@ def get_initial_state(self, inputs):
 
-        is_keras_tensor = hasattr(initial_state[0], '_keras_history')
-        for tensor in initial_state:
+        check_list = []
```
Unclear what is meant by `check_list`. Maybe `input_list` would be better?

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -440,62 +474,68 @@ def get_initial_state(self, inputs):
-                raise ValueError('The initial state of an RNN layer cannot be'
-                                 ' specified with a mix of Keras tensors and'
-                                 ' non-Keras tensors')
+                raise ValueError('The initial state and constants of an RNN'
```
and -> or

andhus replied:
Sure, I think I wanted to refer to that they "jointly" must contain only one kind of tensors (that's what checked for now)... I don't have any opinion on the error msg, but just to verify/understand: shouldn't all inputs be keras tensors or not - and why not include `inputs` then as well in the check (not only `constants` and `initial_state`)?

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -560,6 +614,48 @@ def call(self, inputs, mask=None, training=None, initial_state=None):
             return output
 
+    def _normalize_args(self, inputs, initial_state, constants):
+        """When running a model loaded from file, the input tensors
```
Nit: first line of docstring (one-line summary) should fit in one line and end with a period.

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -560,6 +614,48 @@ def call(self, inputs, mask=None, training=None, initial_state=None):
 
+    def _normalize_args(self, inputs, initial_state, constants):
+        """When running a model loaded from file, the input tensors
+        `initial_state` and `constants` can be passed to RNN.__call__ as part
```
` around code keywords

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -560,6 +614,48 @@ def call(self, inputs, mask=None, training=None, initial_state=None):
+        constants.
+
+        This method makes sure initial_states and constants are separated from
+        inputs and that the are lists of tensors (or None).
```
the -> they

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -560,6 +614,48 @@ def call(self, inputs, mask=None, training=None, initial_state=None):
+                initial_state = inputs[1:]
+            inputs = inputs[0]
+
+        def to_list_or_none(x):  # TODO break out?
```
Unless you want to use it elsewhere, it is fine as it is, you can remove this comment

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -618,6 +714,9 @@ def get_config(self):
                   'stateful': self.stateful,
                   'unroll': self.unroll}
+        if self._n_constants is not None:
+            config['_n_constants'] = self._n_constants
```
Please call it `num_constants` is the config.

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -292,6 +293,14 @@ class RNN(Layer):
 
+    # Note on passing external constants to RNNs
+        You can pass "external" constants to the cell using the `constants`
+        keyword argument of RNN.__call__ (as well as RNN.call) method. This
```
` around code keywords

andhus commented on keras/layers/recurrent.py:
```diff
@@ -415,19 +422,46 @@ def compute_mask(self, inputs, mask):
             return output_mask
 
     def build(self, input_shape):
+        # Note input_shape will be list of shapes of initial states and
```
`will be list of shapes of input followed by shapes of initial states and...`

andhus commented:
I fixed the comments and added a test case for when _both_ `initial_state`and `constants` are passed and tests now also covers "flat list" inputs after serialisation.

What do you think: should we aim to merge this first separately (a pretty well defined feature addition) or continue to pull in next steps (actual attention mechanisms) before merging anything?

fchollet commented:
> What do you think: should we aim to merge this first separately (a pretty well defined feature addition)

We'll merge it first.

andhus commented:
@fchollet saw that build is failing on master: https://travis-ci.org/fchollet/keras/jobs/292414692 `test_plot_model`-> "ImportError: Failed to import pydot..."

dunnock commented:
@andhus it seems after this merged into the master CuDNNLSTM has broke. After pulling I get error:
```
  File "C:\ProgramData\Anaconda3\lib\site-packages\keras-2.0.8-py3.6.egg\keras\layers\recurrent.py", line 482, in __call__
  File "C:\ProgramData\Anaconda3\lib\site-packages\keras-2.0.8-py3.6.egg\keras\engine\topology.py", line 575, in __call__
  File "C:\ProgramData\Anaconda3\lib\site-packages\keras-2.0.8-py3.6.egg\keras\layers\cudnn_recurrent.py", line 400, in build
  File "C:\ProgramData\Anaconda3\lib\site-packages\keras-2.0.8-py3.6.egg\keras\layers\recurrent.py", line 427, in build
AttributeError: 'CuDNNLSTM' object has no attribute '_num_constants'
```
I have latest nightly tensorflow

fchollet commented:
Unfortunately the cuDNN RNN tests cannot be run on Travis since they require a GPU. We need to remember to run them manually everytime we make changes to `recurrent.py`.

I will look at the issue.

andhus commented:
Yeah, thanks for taking care of this, did not think about that addition!

andhus commented:
Continuation of Recurrent Attention API here: https://github.com/fchollet/keras/pull/8296
